torchtitan/train_spec.py

```diff
@@ -0,0 +1,91 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+#
+# Copyright (c) Meta Platforms, Inc. All Rights Reserved.
+
+
+from dataclasses import dataclass
+from typing import Callable, Dict, Protocol, Type, TypeAlias
+
+import torch.nn as nn
+from torch.distributed.pipelining.schedules import _PipelineSchedule
+
+from torchtitan.config_manager import JobConfig
+from torchtitan.optimizer import LRSchedulersContainer, OptimizersContainer
+
+
+@dataclass
+class BaseModelArgs:
+    """All ModelArgs should inherit from this class.
+
+    The only usage of this class is type checking but allows us to extend common
+    arguments to all models in the future.
+    """
+
+    _enforced: str = "This field is used to enforce all fields have defaults."
+
+
+class ModelProtocol(Protocol):
+    """Defines the interface for a model class.
+
+    This is used to enforce that all model classes have some methods that are
+    required by the TorchTitan trainer.
+    """
+
+    @staticmethod
+    def from_model_args(args: BaseModelArgs) -> nn.Module:
+        ...
+
+
+OptimizersBuilder: TypeAlias = Callable[
+    [list[nn.Module], JobConfig], OptimizersContainer
+]
+OptimizerBuilderWrapper: TypeAlias = Callable[
+    [list[nn.Module], JobConfig, OptimizersContainer], OptimizersContainer
+]
+LRSchedulersBuilder: TypeAlias = Callable[[OptimizersContainer], LRSchedulersContainer]
+
+
+@dataclass
+class TrainSpec:
+    name: str
+    cls: Type[nn.Module]
+    config: Dict[str, BaseModelArgs]
+    parallelize_fn: Callable[[nn.Module], None]
+    pipelining_fn: Callable[
+        [nn.Module], tuple[_PipelineSchedule, list[nn.Module], bool, bool]
+    ]
+    build_optimizers_fn: OptimizersBuilder
+    build_lr_schedulers_fn: LRSchedulersBuilder
+
+    # TODO: Add a ``build_dataloader_fn``
+
+    # TODO: Add a FQN convert fn to allow users to load checkpoints from
+    # HuggingFace or other sources that have different FQN conventions.
+
+
+_train_specs = {}
+
+
+def register_train_spec(train_spec: TrainSpec) -> None:
+    global _train_specs
+    if train_spec.name in _train_specs:
+        raise ValueError(f"Model {train_spec.name} is already registered.")
+
+    _train_specs[train_spec.name] = train_spec
+
+
+def get_train_spec(name: str) -> TrainSpec:
+    global _train_specs
+    if name not in _train_specs:
+        raise ValueError(f"Model {name} is not registered.")
+    return _train_specs[name]
+
+
+def apply_to_train_specs(func: Callable[[TrainSpec], TrainSpec]) -> None:
```

Comment:
IIUC, there are two ways to change the model Spec. One, user define a brand new model spec so that every part is customized. Two, user use this function to swap some part (optimizer) and still reuse existing model config?

Reply:
Synced offline, turns out this is for hook registration.